examples/energy-management-app/energy-management-common/src/ChargingTargetsMemMgr.cpp

```diff
@@ -0,0 +1,154 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app-common/zap-generated/cluster-objects.h>
+
+#include "ChargingTargetsMemMgr.h"
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::DataModel;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::EnergyEvse;
+
+ChargingTargetsMemMgr::ChargingTargetsMemMgr() : mChargingTargetSchedulesIdx(0), mNumDailyChargingTargets(0)
+{
+    memset(mpListOfDays, 0, sizeof(mpListOfDays));
+}
+
+ChargingTargetsMemMgr::~ChargingTargetsMemMgr()
+{
+    // Free all memory allocated for the charging targets
+    for (uint16_t idx = 0; idx < kEvseTargetsMaxNumberOfDays; idx++)
+    {
+        if (mpListOfDays[idx] != nullptr)
+        {
+            chip::Platform::Delete(mpListOfDays[idx]);
+        }
+    }
+}
+
+void ChargingTargetsMemMgr::Reset(uint16_t chargingTargetSchedulesIdx)
+{
+    // MUST be called for each entry in DataModel::List<const Structs::ChargingTargetScheduleStruct::Type> chargingTargetSchedules
+    mNumDailyChargingTargets    = 0;
+    mChargingTargetSchedulesIdx = chargingTargetSchedulesIdx;
+
+    // Free up any memory associated with this targetSchedule
+    if (mpListOfDays[mChargingTargetSchedulesIdx] != nullptr)
+    {
+        chip::Platform::MemoryFree(mpListOfDays[mChargingTargetSchedulesIdx]);
+        mpListOfDays[mChargingTargetSchedulesIdx] = nullptr;
+    }
+}
+
+void ChargingTargetsMemMgr::AddChargingTarget(EnergyEvse::Structs::ChargingTargetStruct::Type & chargingTarget)
+{
+    if (mNumDailyChargingTargets < kEvseTargetsMaxTargetsPerDay)
+    {
+        mDailyChargingTargets[mNumDailyChargingTargets++] = chargingTarget;
+    }
+}
+
+EnergyEvse::Structs::ChargingTargetStruct::Type * ChargingTargetsMemMgr::GetChargingTargets() const
+{
+    return mpListOfDays[mChargingTargetSchedulesIdx];
+}
+
+uint16_t ChargingTargetsMemMgr::GetNumDailyChargingTargets() const
+{
+    return mNumDailyChargingTargets;
+}
+
+void ChargingTargetsMemMgr::AllocAndCopy()
+{
+    if (mNumDailyChargingTargets > 0)
+    {
+        // Allocate the memory first and then use placement new to initialise the memory of each element in the array
+        mpListOfDays[mChargingTargetSchedulesIdx] = static_cast<EnergyEvse::Structs::ChargingTargetStruct::Type *>(
+            chip::Platform::MemoryAlloc(sizeof(EnergyEvse::Structs::ChargingTargetStruct::Type) * mNumDailyChargingTargets));
```

Comment:
should you check if allocation succeeded/you had sufficient RAM? Or is segfault acceptable?

If we check, maybe AllocAndCopy should return a CHIP_ERROR.

Reply:
A check is now made and CHIP_ERROR returned.